### PR TITLE
Jacklanchantin/online training

### PR DESCRIPTION
**Pull request by jacklanchantin**
**What does this PR do? Please describe:**
A summary of the change or the issue that is fixed.

Introduces using a vllm-based reward model to score (prompt, rollout) pairs. This requires a second RemoteVllmModel in addition to the rollout RemoteVllmModel. 

**Does your PR introduce any breaking changes? If yes, please list them:**
It requires changes to the creators of VLLMOutputRewardHandler so that they all accept a reward_model (which can be the vllm RM pointer).

**Check list:**
- [ ] Was the content of this PR **discussed and approved** via a GitHub issue? (no need for typos or documentation improvements)
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairseq2/blob/main/CONTRIBUTING.md)?
- [ ] Did you make sure that your **PR does only one thing** instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**?
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you **update the [CHANGELOG](https://github.com/facebookresearch/fairseq2/blob/main/CHANGELOG.md)**? (no need for typos, documentation, or minor internal changes)

### CODE REVIEW

File: src/fairseq2/recipes/lm/_online_finetune/_common.py [hunk 1]
```diff
@@ -255,47 +272,63 @@ def copy_state(src_module: nn.Module, tgt_module: nn.Module):
...
 
-        rollouts_to_scatter = []
-        rollouts_per_rank = [None]
+        if operation == "rollout":
+            outputs = vllm_model.rollout_from_model(
+                flat_request_list, sampling_params=sampling_params
+            )
+
+        elif operation == "reward":
+            outputs = vllm_model.reward_from_model(flat_request_list)
```
Comment (jacklanchantin) [inline src/fairseq2/recipes/lm/_online_finetune/_common.py:321]: I wanted to keep both of these within the same `generate_rollouts()`, but not sure if this is the best way to do it.

File: src/fairseq2/recipes/lm/_online_finetune/_common.py [hunk 2]
```diff
@@ -302,23 +301,33 @@ def find_first_value(lst, value):
...
-        rollouts = vllm_model.rollout_from_model(
-            flat_request_list, sampling_params=sampling_params
-        )
+        if operation == "rollout":
+            rollouts = vllm_model.rollout_from_model(
+                flat_request_list, sampling_params=sampling_params
+            )
+
+        elif operation == "reward":
+            rollouts = vllm_model.reward_from_model(flat_request_list)
```
Comment (jacklanchantin) [inline src/fairseq2/recipes/lm/_online_finetune/_common.py:330]: I wanted to keep both of these calls in the same function generate_rollouts() function, but not sure if this is the best way to do it.

File: src/fairseq2/recipes/lm/_online_finetune/_remote_vllm.py [hunk 3]
```diff
@@ -175,6 +179,9 @@ def setup_vllm_worker(
             enforce_eager=vllm_engine_args.enforce_eager,
             worker_cls=MyWorker,
             tensor_parallel_size=vllm_engine_args.tensor_parallel_size,
+            task=vllm_engine_args.task,
+            hf_overrides=vllm_engine_args.hf_overrides,
+            override_pooler_config=vllm_engine_args.override_pooler_config,
```
Comment (jacklanchantin) [inline src/fairseq2/recipes/lm/_online_finetune/_remote_vllm.py:184]: these are required for some reward models in vllm

File: src/fairseq2/recipes/lm/_online_finetune/_rewards.py [hunk 4]
```diff
@@ -77,7 +80,7 @@ def __init__(self):
         pass
 
     @override
-    def create(self, reward_config, gangs):
+    def create(self, reward_model, reward_config, gangs):
```
Comment (jacklanchantin) [inline src/fairseq2/recipes/lm/_online_finetune/_rewards.py:83]: added this so that all reward_handlers accept this by default

File: src/fairseq2/recipes/lm/_online_finetune/_grpo.py [hunk 5]
```diff
@@ -428,6 +432,7 @@ class GrpoFinetuneConfig:
     """The data type of the reference model."""
 
     ray_policy_actor_name: str = "vllm_policy"
+    vllm_reward_model_name: str = None
```
Comment (jacklanchantin) [inline src/fairseq2/recipes/lm/_online_finetune/_grpo.py:435]: will be None by default, for rewards like GSM8K verifier

File: src/fairseq2/recipes/lm/_online_finetune/_common.py [hunk 6]
```diff
@@ -302,23 +295,33 @@ def find_first_value(lst, value):
 
 
 def generate_rollouts(
-    prompts: List[List[int]], dp_gang: Gang, vllm_model, sampling_params=None
+    prompts: List[List[int]],
+    dp_gang,
+    vllm_model,
+    operation: str = "rollout",
+    sampling_params=None,
```
Comment (jacklanchantin) [inline src/fairseq2/recipes/lm/_online_finetune/_common.py:302]: formatting

File: src/fairseq2/recipes/lm/_online_finetune/_common.py [hunk 7]
```diff
@@ -302,23 +295,33 @@ def find_first_value(lst, value):
...
 ):
+    if operation not in ["rollout", "reward"]:
+        raise ValueError("Operation must be either 'rollout' or 'reward'.")
+
```
Comment (jacklanchantin) [inline src/fairseq2/recipes/lm/_online_finetune/_common.py:306]: see comment below on whether we should do it this way

File: src/fairseq2/recipes/lm/_online_finetune/_grpo.py [hunk 8]
```diff
@@ -29,49 +31,44 @@
...
+    RewardSection,
+    VLLMOutputReward,
+    VLLMOutputRewardHandler,
 )
 from fairseq2.recipes.metrics import SequenceMetricBag
+from fairseq2.recipes.model import Model
+from fairseq2.recipes.trainer import TrainUnit
+from fairseq2.typing import DataType
+from fairseq2.utils.structured import structure
+from fairseq2.utils.validation import validate
```
Comment (jacklanchantin) [inline src/fairseq2/recipes/lm/_online_finetune/_grpo.py:71]: from isort

File: src/fairseq2/recipes/lm/_online_finetune/_grpo.py [hunk 9]
```diff
@@ -167,7 +164,9 @@ def validate_reward(self, prompt_batch: PromptBatch) -> tuple[Tensor, int]:
             sampling_params=policy_sampling_params,
         )
         reward_output = self._reward.process_rollouts(
-            rollouts, prompt_batch.meta_info[self._reward.answer_key]
+            rollouts,
+            prompt_batch.meta_info[self._reward.answer_key],
+            prompt_batch.meta_info[self._reward.prompt_key],
```
Comment (jacklanchantin) [inline src/fairseq2/recipes/lm/_online_finetune/_grpo.py:169]: need raw prompts for skywork, so passing them in here

File: src/fairseq2/recipes/lm/_online_finetune/_handler.py [hunk 10]
```diff
@@ -10,24 +10,27 @@
...
     @abstractmethod
-    def name(self) -> str: ...
+    def name(self) -> str:
+        ...
 
     @property
     @abstractmethod
-    def config_kls(self) -> type[object]: ...
+    def config_kls(self) -> type[object]:
+        ...
```
Comment (jacklanchantin) [inline src/fairseq2/recipes/lm/_online_finetune/_handler.py:33]: formatting

File: src/fairseq2/recipes/lm/_online_finetune/_online_dpo.py [hunk 11]
```diff
@@ -29,53 +33,46 @@
...
-    RemoteVllmModel,
-)
-
-from fairseq2.recipes.lm._online_finetune._common import (
-    collate_with_target_mask,
-    copy_state,
-    find_first_value,
-    generate_rollouts,
-    convert_vllm_output_to_ref_score,
-)
```
Comment (jacklanchantin) [inline src/fairseq2/recipes/lm/_online_finetune/_online_dpo.py:78]: isort

Comment (jacklanchantin) [inline src/fairseq2/recipes/lm/_online_finetune/_online_dpo.py:78]: isort/black

File: src/fairseq2/recipes/lm/_online_finetune/_recipe.py [hunk 12]
```diff
@@ -16,6 +16,11 @@
```
Comment (jacklanchantin) [inline src/fairseq2/recipes/lm/_online_finetune/_recipe.py:16]: all isort/black

File: src/fairseq2/recipes/lm/_online_finetune/_remote_vllm.py [hunk 13]
```diff
@@ -9,47 +9,52 @@
...
 
 @dataclass(kw_only=True)
 class VllmEngineArgs:
     model: str = "/checkpoint/ram/kulikov/gsm8k_8b_sft/checkpoints/step_20"
     tokenizer: str = "/datasets/pretrained-llms/Llama-3.1-8B-Instruct"
+    task: str = "generate"
     tensor_parallel_size: int = 4
     enforce_eager: bool = True
+    hf_overrides: object = None
+    override_pooler_config: PoolerConfig = field(default_factory=lambda: PoolerConfig())
```
Comment (jacklanchantin) [inline src/fairseq2/recipes/lm/_online_finetune/_remote_vllm.py:57]: these are required for some reward models in vllm

File: src/fairseq2/recipes/lm/_online_finetune/_rewards.py [hunk 14]
```diff
@@ -181,14 +192,238 @@ def config_kls(self):
...
+        )
+
+    def extract_answer(self, completion: str):
+        match = self.answer_re.search(completion)
+        if match:
+            match_str = match.group(1).strip()
+            match_str = match_str.replace(",", "")
+            return match_str
+        else:
+            return self.invalid_answer
```
Comment (uralik) [inline src/fairseq2/recipes/lm/_online_finetune/_rewards.py:254]:
```suggestion
```

**Review by uralik — APPROVED** [review]
thanks!!
